File: iOS-Native/MentalHealthCompanion/Services/CommunityService.swift
import Foundation
import Combine
// MARK: - Community Service (matching community_service.dart)
class CommunityService: ObservableObject {

    private let baseURL = "https://gen-ai-g6tt.onrender.com/api/v1/community"

    func createCommunityMessage(userId: String, message: String) async {
        guard let url = URL(string: "\(baseURL)/create") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["userId": userId, "message": message]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                print("Message created")
            } else {
                print("Failed to create message")
            }
        } catch {
            print("Error creating messages: \(error)")
        }
    }

    func addComment(userId: String, comment: String) async {
        guard let url = URL(string: "\(baseURL)/addComment/\(userId)") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["userId": userId, "comment": comment]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                print("Comment sent")
            } else {
                print("Failed to send comment")
            }
        } catch {
            print("Error sending comment: \(error)")
        }
    }

    func getCommunityMessages() async -> [[String: Any]] {
        guard let url = URL(string: "\(baseURL)/get") else { return [] }

        var request = URLRequest(url: url)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200,
               let messages = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
                print("Got messages")
                return messages
            }
            print("Failed to get message")
            return []
        } catch {
            print("Error getting messages: \(error)")
            return []
        }
    }

    func getCommunityMessagesById(userId: String) async -> [[String: Any]] {
        guard let url = URL(string: "\(baseURL)/get/\(userId)") else { return [] }

        var request = URLRequest(url: url)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200,
               let messages = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] {
                print("Got messages")
                return messages
            }
            print("Failed to get message")
            return []
        } catch {
            print("Error getting messages: \(error)")
            return []
        }
    }
}

File: app/lib/services/community_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:saathi/services/api_helpers.dart';

class CommunityService {
  Future<void> createCommunityMessage(String userId, String message) async {
    try {
      final url =
          Uri.parse('https://gen-ai-g6tt.onrender.com/api/v1/community/create');
      final headers = {'Content-Type': 'application/json'};
      final body = jsonEncode({
        'userId': userId,
        'message': message,
      });
      final response = await http.post(url, headers: headers, body: body);
      if (response.statusCode == 200) {
        print('Message created');
      } else {
        throw BackendException('Failed to create message', statusCode: response.statusCode);
      }
    } catch (e) {
      print("Error creating messages: $e");
      rethrow;
    }
  }

  Future<void> addComment(String userId, String comment) async {
    try {
      final url = Uri.parse(
          'https://gen-ai-g6tt.onrender.com/api/v1/community/addComment/$userId');
      final headers = {'Content-Type': 'application/json'};
      final body = jsonEncode({
        'userId': userId,
        'comment': comment,
      });
      final response = await http.post(url, headers: headers, body: body);
      if (response.statusCode == 200) {
        print('comment sent');
      } else {
        throw BackendException('Failed to send comment', statusCode: response.statusCode);
      }
    } catch (e) {
      print("Error sending comment: $e");
      rethrow;
    }
  }

  Future<List<dynamic>> getCommunityMessages() async {
    return retryWithBackoff(() async {
      final url =
          Uri.parse('https://gen-ai-g6tt.onrender.com/api/v1/community/get');
      final headers = {'Content-Type': 'application/json'};
      final response = await http.get(url, headers: headers);
      if (response.statusCode == 200) {
        final messages = safeJsonDecode(response);
        if (messages is List) {
          return messages;
        } else {
          throw BackendException('Invalid community messages format', statusCode: response.statusCode);
        }
      } else {
        throw BackendException('Failed to get message', statusCode: response.statusCode);
      }
    });
  }

  Future<List<dynamic>> getCommunityMessagesById(String userId) async {
    return retryWithBackoff(() async {
      final url = Uri.parse(
          'https://gen-ai-g6tt.onrender.com/api/v1/community/get/$userId');
      final headers = {'Content-Type': 'application/json'};
      final response = await http.get(url, headers: headers);
      if (response.statusCode == 200) {
        final messages = safeJsonDecode(response);
        if (messages is List) {
          return messages;
        } else {
          throw BackendException('Invalid community messages format', statusCode: response.statusCode);
        }
      } else {
        throw BackendException('Failed to get message', statusCode: response.statusCode);
      }
    });
  }

  Future<void> reportPost(String postId, String reason, String details) async {
    try {
      await FirebaseFirestore.instance.collection('reports').add({
        'postId': postId,
        'reason': reason,
        'details': details,
        'timestamp': FieldValue.serverTimestamp(),
      });
      print('Post reported successfully');
    } catch (e) {
      print('Error reporting post: $e');
      rethrow;
    }
  }
}
